### ISSUE TITLE
BehaviourSubject toPromise isn't working

### ISSUE THREAD

**Issue by pjpsoares** [Reactive-Extensions/RxJS#1088]
It seems that the promise returned by behaviourSubject never actually get resolved.
Sample code:
```
let bs = new Rx.BehaviorSubject('some value');
bs.toPromise().then((value) => console.log(value));
```

**Comment by mattpodwysocki** [comment]
@pjpsoares in order to convert it to a promise, it must have an `onNext` call followed by a `onCompleted` call, which it doesn't have.  Is the desired `toPromise` behavior to take the first value and then do `onNext` and `onCompleted`?

**Comment by pjpsoares** [comment]
Yup, that was what I was expecting. It feels kind of awkward the promise only being resolved with the last value. With the current value seems more appropriate, but still a little strange. Honestly, when I think of it I just think it shouldn't support the toPromise at all...

**Comment by jadbox** [comment]
@mattpodwysocki Doesn't the promise spec allow for a promise to be resolved without a value: e.g. Bluebird's Promise.resolve() with not pass a value to 'then' handlers as it has no value [it's not a null value].

**Comment by paulpdaniels** [comment]
@jadbox that would still work as expected. `toPromise` just calls `resolve` under the hood, so if the Observable just completed then it would call `resolve(undefined)`

**Comment by NoNameProvided** [comment]
Just experienced the same "bug". I had expected the current value stored in the `BehaviourSubject` as well. Is this open for discussion @mattpodwysocki? Do we have any built in way to get a (Behaviour)Subject value async, only once, without the need of handling the subscription? 

In the meantime if anyone wondering how get the current value without leaking subscriptions here is an example. (TypeScript code)

```ts
public geValuesAsPromise() : Promise<IValue[]> {
    let resolver = (subsription: Subscription, resolve: Function, result: IValue[]) => { resolve(result); subsription.unsubscribe(); };
    let rejecter = (subsription: Subscription, reject: Function, error: any) => { reject(error); subsription.unsubscribe(); };

    return new Promise((resolve, reject) => {
      let subscription = this.getValueObservable().subscribe( 
        result => { resolver(subscription, resolve, result); }, 
        error => { rejecter(subscription, reject, error); }
        );
    });
  }
```

**Comment by gforceg** [comment]
I ran into this same issue just now. `toPromise().then( obj ...` passes in 'undefined' for obj.
I just wanted to lazy cache data from a XHR request into a service in my app.
Is this a bug or should I be using real promises every time I interact with my service?

**Comment by zixia** [comment]
I had expected the current value stored in the BehaviourSubject as well too. 

I believe that's the reason why we design the BehaviorSubject right? Always get the last value or the initial value.

Strongly suggest we should consider to add this feature, because it's very natural in our mind to be like this: return the last value or the initial value.

**Comment by paulpdaniels** [comment]
For anyone here who is still interested in this issue, I would suggest raising it over at the https://github.com/ReactiveX/rxjs repo, because this kind of behavior change will definitely break things and will *not* make it into this library which is in maintenance mode.

In all likelihood, I doubt that this behavior will change since it is trivial to implement this with existing operators.

```
const behaviorSubject = new Rx.BehaviorSubject(42);

behaviorSubject.first().toPromise().then(function(x) {
  console.log("The first answer is " + x);
})

behaviorSubject.take(1).toPromise().then(function(x) {
  console.log("The second answer is " + x);
})

```

**Comment by zixia** [comment]
Thanks, @paulpdaniels

However, I would like to suggest this kind of behavior change is not a breaking change but a **bug fix**. ;-)

**Comment by NoNameProvided** [comment]
If it's change the current behaviour then it's a breaking change not a bugfix.

**Comment by zixia** [comment]
@NoNameProvided Sorry I can't agree about this.

Because every bugfix will change the current behavior of the code, isn't it? ;-)

I believe what we care about is not about the _change_, but whether it is **expected**.

**Comment by NoNameProvided** [comment]
No, a bugfix doesnt change the behavior, it makes the behavior consistent.

**Comment by paulpdaniels** [comment]
@zixia The example above is not buggy behavior, it is intentional. It occurs because the compatibility between `Promises` and `Observables` is inherently messy because you are moving from a more expressive type to a less expressive one. `Observables` have both `onNext` and `onComplete` because there can be multiple events. Promises only have `resolve` because completion is implicit. This means that there has to be some form of explicit completion (as outlined above through the use of `take(1)` or `first()`) in order to properly map the semantics onto each other.

**Comment by huan** [comment]
@paulpdaniels Thanks for your explanation.

I fully agree with you about the different between the Promise and Observable.

~~However, from the view of the Observable, I still feel some behavior is not intentional(in natural). Let's start from comparing the behavior between BehaviorSubject and the Observable:~~

```ts
const bs = new BehaviorSubject(42)
bs.subscribe(n => console.log)
// output 42
```
v.s.

```ts
const os = bs.asObservable()
os.subscribe(n => console.log)
// ~~nothing emitted at here~~
// Update: here emit init value as well.
```

As my understanding: a BehaviorSubject is also an Observable, that's why we could subscribe to it.

So here's what I feel uncomfortable:

1. subscribe to Observable A will get an initial value
1. ~~subscribe to Observable B will NOT get initial value.~~
1. however: `B = A.asObservable()`

Update: **Sorry, after double check, I found that `bs.asObservable().subscribe(n => console.log)` work with the init value as well**.

**Comment by piernik** [comment]
In my opinion it's a bug. `toPromise` should be immediately resolved if Observable has value. 
I've spend over two hours to reach this page to find out that this `bug` is a `feature` :/
